File: watchos/Models/Presets.swift
// Component presets — Unicode box-drawing wireframe templates.
//
// Verbatim copy of ios/App/Presets.swift (itself transcribed from src/lib/presets.js) so all
// three surfaces keep the same 23 components. Multi-row/wide templates like Modal or Browser
// simply clip against the smaller watch grid, same as any oversized stamp clips on iOS —
// no special-casing needed.

import Foundation

struct Preset: Identifiable, Hashable {
    let id: String
    let label: String
    let category: String
    let template: [String]

    var height: Int { template.count }
    var width: Int { template.map(\.count).max() ?? 0 }
}

let presets: [Preset] = [
    Preset(id: "button", label: "Button", category: "Input", template: ["[ OK ]"]),
    Preset(id: "input", label: "Input", category: "Input", template: ["[____________]"]),
    Preset(id: "select", label: "Select", category: "Input", template: ["[▾           ]"]),
    Preset(id: "checkbox", label: "Checkbox", category: "Input", template: ["[✓] Label"]),
    Preset(id: "radio", label: "Radio", category: "Input", template: ["(●) Option"]),
    Preset(id: "toggle", label: "Toggle", category: "Input", template: ["[●━━━━━]"]),
    Preset(id: "table", label: "Table", category: "Data", template: [
        "┌──────┬──────┬──────┐",
        "│ Head │ Head │ Head │",
        "├──────┼──────┼──────┤",
        "│ Cell │ Cell │ Cell │",
        "│ Cell │ Cell │ Cell │",
        "└──────┴──────┴──────┘",
    ]),
    Preset(id: "modal", label: "Modal", category: "Overlay", template: [
        "┌────────────[×]──────┐",
        "│ Modal Title         │",
        "│                     │",
        "│ Content goes here   │",
        "│                     │",
        "│         [ Cancel ]  [ OK ] │",
        "└─────────────────────┘",
    ]),
    Preset(id: "browser", label: "Browser", category: "Layout", template: [
        "┌─────────────────────────────┐",
        "│ ◄ ► ⟳  [___________________]│",
        "├─────────────────────────────┤",
        "│                             │",
        "│                             │",
        "│                             │",
        "└─────────────────────────────┘",
    ]),
    Preset(id: "card", label: "Card", category: "Layout", template: [
        "┌──────────────────┐",
        "│ Card Title       │",
        "│                  │",
        "│ Card content     │",
        "│ goes here.       │",
        "│                  │",
        "│ [ Action ]       │",
        "└──────────────────┘",
    ]),
    Preset(id: "navbar", label: "Navbar", category: "Layout", template: [
        "┌─────────────────────────────────┐",
        "│ ≡  Logo            Search    ☆  │",
        "└─────────────────────────────────┘",
    ]),
    Preset(id: "tabs", label: "Tabs", category: "Layout", template: [
        "┌──────┐  ────────  ────────",
        "│ Tab1 │   Tab 2     Tab 3  ",
        "├──────┴──────────────────┐",
        "│ Content                  │",
        "└──────────────────────────┘",
    ]),
    Preset(id: "progress", label: "Progress", category: "Feedback", template: ["▓▓▓▓▓▓░░░░░░░░ 42%"]),
    Preset(id: "icon", label: "Icon", category: "Media", template: ["★"]),
    Preset(id: "image", label: "Image", category: "Media", template: [
        "╔══════════════╗",
        "║              ║",
        "║   [ img ]    ║",
        "║              ║",
        "╚══════════════╝",
    ]),
    Preset(id: "divider", label: "Divider", category: "Layout", template: ["─────────────────────"]),
    Preset(id: "alert", label: "Alert", category: "Feedback", template: [
        "┌──────────────────────────┐",
        "│ ⚠  Warning message here  │",
        "└──────────────────────────┘",
    ]),
    Preset(id: "breadcrumb", label: "Breadcrumb", category: "Navigation", template: ["Home > Section > Page"]),
    Preset(id: "avatar", label: "Avatar", category: "Media", template: [
        "╭──╮",
        "│DH│",
        "╰──╯",
    ]),
    Preset(id: "list", label: "List", category: "Data", template: [
        "• Item one",
        "• Item two",
        "• Item three",
    ]),
    Preset(id: "stepper", label: "Stepper", category: "Navigation", template: ["● ─────── ○ ─────── ○"]),
    Preset(id: "rating", label: "Rating", category: "Feedback", template: ["★★★☆☆"]),
    Preset(id: "skeleton", label: "Skeleton", category: "Feedback", template: [
        "░░░░░░░░░░░░░░░░░░░░",
        "░░░░░░░░░░░░░░",
        "░░░░░░░░░░░░░░░░░",
    ]),
]

/// Categories in first-appearance order, matching the JS `[...new Set(...)]`.
let presetCategories: [String] = presets.reduce(into: [String]()) { acc, p in
    if !acc.contains(p.category) { acc.append(p.category) }
}

func preset(id: String) -> Preset? {
    presets.first { $0.id == id }
}

File: kmp/shared/src/commonMain/kotlin/com/nulljosh/charwork/Presets.kt
package com.nulljosh.charwork

// Ported from src/lib/presets.js / ios/App/Presets.swift. Keep line-comparable.
data class Preset(val id: String, val label: String, val category: String, val template: List<String>, val width: Int, val height: Int)

val PRESETS: List<Preset> = listOf(
    Preset("button", "Button", "Input", listOf("[ OK ]"), 6, 1),
    Preset("input", "Input", "Input", listOf("[____________]"), 14, 1),
    Preset("select", "Select", "Input", listOf("[▾           ]"), 14, 1),
    Preset("checkbox", "Checkbox", "Input", listOf("[✓] Label"), 9, 1),
    Preset("radio", "Radio", "Input", listOf("(●) Option"), 10, 1),
    Preset("toggle", "Toggle", "Input", listOf("[●━━━━━]"), 8, 1),
    Preset("table", "Table", "Data", listOf("┌──────┬──────┬──────┐", "│ Head │ Head │ Head │", "├──────┼──────┼──────┤", "│ Cell │ Cell │ Cell │", "│ Cell │ Cell │ Cell │", "└──────┴──────┴──────┘"), 22, 6),
    Preset("modal", "Modal", "Overlay", listOf("┌────────────[×]──────┐", "│ Modal Title         │", "│                     │", "│ Content goes here   │", "│                     │", "│         [ Cancel ]  [ OK ] │", "└─────────────────────┘"), 25, 7),
    Preset("browser", "Browser", "Layout", listOf("┌─────────────────────────────┐", "│ ◄ ► ⟳  [___________________]│", "├─────────────────────────────┤", "│                             │", "│                             │", "│                             │", "└─────────────────────────────┘"), 33, 7),
    Preset("card", "Card", "Layout", listOf("┌──────────────────┐", "│ Card Title       │", "│                  │", "│ Card content     │", "│ goes here.       │", "│                  │", "│ [ Action ]       │", "└──────────────────┘"), 20, 8),
    Preset("navbar", "Navbar", "Layout", listOf("┌─────────────────────────────────┐", "│ ≡  Logo            Search    ☆  │", "└─────────────────────────────────┘"), 35, 3),
    Preset("tabs", "Tabs", "Layout", listOf("┌──────┐  ────────  ────────", "│ Tab1 │   Tab 2     Tab 3  ", "├──────┴──────────────────┐", "│ Content                  │", "└──────────────────────────┘"), 30, 5),
    Preset("progress", "Progress", "Feedback", listOf("▓▓▓▓▓▓░░░░░░░░ 42%"), 18, 1),
    Preset("icon", "Icon", "Media", listOf("★"), 1, 1),
    Preset("image", "Image", "Media", listOf("╔══════════════╗", "║              ║", "║   [ img ]    ║", "║              ║", "╚══════════════╝"), 16, 5),
    Preset("divider", "Divider", "Layout", listOf("─────────────────────"), 21, 1),
    Preset("alert", "Alert", "Feedback", listOf("┌──────────────────────────┐", "│ ⚠  Warning message here  │", "└──────────────────────────┘"), 28, 3),
    Preset("breadcrumb", "Breadcrumb", "Navigation", listOf("Home > Section > Page"), 21, 1),
    Preset("avatar", "Avatar", "Media", listOf("╭──╮", "│DH│", "╰──╯"), 4, 3),
    Preset("list", "List", "Data", listOf("• Item one", "• Item two", "• Item three"), 12, 3),
    Preset("stepper", "Stepper", "Navigation", listOf("● ─────── ○ ─────── ○"), 21, 1),
    Preset("rating", "Rating", "Feedback", listOf("★★★☆☆"), 5, 1),
    Preset("skeleton", "Skeleton", "Feedback", listOf("░░░░░░░░░░░░░░░░░░░░", "░░░░░░░░░░░░░░", "░░░░░░░░░░░░░░░░░"), 20, 3),
)

val CATEGORIES: List<String> = PRESETS.map { it.category }.distinct()

fun getPreset(id: String): Preset? = PRESETS.find { it.id == id }
